### PR TITLE
Fix: moved to prewhere condition actions can lose column

### LINKED ISSUE #53237: move_all_conditions_to_prewhere / NOT_FOUND_COLUMN_IN_BLOCK
https://fiddle.clickhouse.com/e7d218c3-2d5f-4a7b-b713-0b5052b62e86

```sql
CREATE TABLE  t  ( e String, c String, q String )
ENGINE = MergeTree ORDER BY tuple();

insert into t
select number, number, number from numbers(10);

WITH s AS ( SELECT * FROM  t WHERE ((e != 'cl') OR (q = 'bn')))
SELECT count() AS cnt
FROM s WHERE (q = 'bn')  GROUP BY c ORDER BY cnt DESC;



Received exception from server (version 23.7.3):
Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column equals(q, 'bn') in block: 
while executing 'INPUT : 1 -> equals(q, 'bn') UInt8 : 2': While executing MergeTreeInOrder. (NOT_FOUND_COLUMN_IN_BLOCK)
(query: WITH s AS ( SELECT * FROM  t WHERE ((e != 'cl') OR (q = 'bn')))
SELECT count() AS cnt
FROM s WHERE (q = 'bn')  GROUP BY c ORDER BY cnt DESC
format Pretty;)
```

worked before 23.7
works with analyzer

-----
**upd: move_all_conditions_to_prewhere**

```sql
CREATE TABLE  t  ( e String, c String, q String ) ENGINE = MergeTree ORDER BY tuple();

insert into t select number, number, number from numbers(10);

SELECT count() 
FROM (SELECT * FROM t WHERE e = 'cl' OR q = 'bn') WHERE (q = 'bn')  
GROUP BY c

DB::Exception: Not found column equals(q, 'bn') in block: while executing 'INPUT : 1 -> equals(q, 'bn') UInt8 : 2': While executing MergeTreeInOrder. (NOT_FOUND_COLUMN_IN_BLOCK)

set move_all_conditions_to_prewhere=0;

SELECT count() 
FROM (SELECT * FROM t WHERE e = 'cl' OR q = 'bn') WHERE (q = 'bn')  
GROUP BY c
0 rows in set. Elapsed: 0.003 sec.
```

### LINKED ISSUE #53453: "DB::Exception: Not found column" when using PREWHERE optimization in SELECT queries
**Describe what's wrong**

Potential problem with [PREWHERE](https://clickhouse.com/docs/en/sql-reference/statements/select/prewhere) optimization in [SELECT](https://clickhouse.com/docs/en/sql-reference/statements/select) queries.

Example: If I have a query where a column is in a NOT clause with the same value as another condition it produces an error like: 

`
Code: 10. DB::Exception: Not found column equals(user_id, 101) in block: while executing 'INPUT : 0 -> equals(user_id, 101) UInt8 : 0': While executing MergeTreeInOrder. (NOT_FOUND_COLUMN_IN_BLOCK) (version 23.8.1.41458 (official build))
`

**Does it reproduce on recent release?**

I'm on `23.8.1.41458`

**How to reproduce**

Following the quick start tutorial (https://clickhouse.com/docs/en/getting-started/quick-start#4-create-a-table)

1). CREATE TABLE: 

```
CREATE TABLE my_first_table
(
    user_id UInt32,
    message String,
    timestamp DateTime,
    metric Float32
)
ENGINE = MergeTree
PRIMARY KEY (user_id, timestamp);
```

2). INSERT data into it: 

```
INSERT INTO my_first_table (user_id, message, timestamp, metric) VALUES
    (101, 'Hello, ClickHouse!',                                 now(),       -1.0    ),
    (102, 'Insert a lot of rows per batch',                     yesterday(), 1.41421 ),
    (102, 'Sort your data based on your commonly-used queries', today(),     2.718   ),
    (101, 'Granules are the smallest chunks of data read',      now() + 5,   3.14159 );
```

3). Run a query: 

```
SELECT *
FROM my_first_table WHERE user_id = 101
AND NOT (user_id = 101 AND (metric = -1.0));
```

gives me the error: 

```
Code: 10. DB::Exception: Not found column equals(user_id, 101) in block: while executing 'INPUT : 0 -> equals(user_id, 101) UInt8 : 0': While executing MergeTreeInOrder. (NOT_FOUND_COLUMN_IN_BLOCK) (version 23.8.1.41458 (official build))
```

4). If you set `optimize_move_to_prewhere = 0` it works as expected: 

```
SELECT *
FROM my_first_table WHERE user_id = 101
AND NOT (user_id = 101 AND (metric = -1.0))
SETTINGS optimize_move_to_prewhere = 0;
```

**Additional context**

Looks similar to https://github.com/ClickHouse/ClickHouse/issues/37381

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed issue when during prewhere optimization compound condition actions DAG can lose output column of intermediate step while this column is required as an input column of some next step.

closes #53453
closes #53237

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit 154b3a1ae4867329206ac6a5d5ef12dc1e9d1685 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53492/154b3a1ae4867329206ac6a5d5ef12dc1e9d1685/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/MergeTree/MergeTreeSplitPrewhereIntoReadSteps.cpp [hunk 1]
```diff
@@ -285,13 +287,13 @@ bool tryBuildPrewhereSteps(PrewhereInfoPtr prewhere_info, const ExpressionAction
         }
 
         steps.push_back({step_dag, result_name});
+        all_output_names.insert(result_name);
```
Comment (novikd) [inline src/Storages/MergeTree/MergeTreeSplitPrewhereIntoReadSteps.cpp:290]: Why are they not in the outputs?

**Comment by novikd** [comment]
@yakov-olkhovskiy I and @davenger modified a bit the fix in regard to allowing the removal of the expressions when they are not used anymore.

**Comment by den-crane** [comment]
turned out it's pretty frequent issue. I would prioritize it.

**Comment by yakov-olkhovskiy** [comment]
@novikd let's finish this PR

**Comment by novikd** [comment]
@yakov-olkhovskiy I suppose it's enough to just update the branch because test failures are unrelated. Let's check out if it helps.

**Pull request merged**